### PR TITLE
4th PR - RecyclerView and Glide implementation

### PR DESCRIPTION
**Pull request by AndresConfortiGL**
## Description
4th PR - RecyclerView and Glide were implemented to the project.

## Story Type
- New feature :white_check_mark:
- Change request
- Bug fix
- Tech

## How Can This Be Tested?
By running the app, you will be able to press the button and get prompted by a loading icon, then the data is fetched and is shown in the RecyclerView.

## Evidences
| Pixel 3a | Nexus S |
| ------------- | ------------- |
|pixel|nexus|

| Coverage Test  |
| ------------- |
|Coverage|

## Checklist
- All unit Test passed :white_check_mark:

## Tested on
The project was tested on Android 8.0 / API 26 using the following devices:

- Pixel 3a
- Nexus S

## Package Structure
| Structure |
| ------------- |
|package|

### CODE REVIEW

File: app/src/main/java/com/example/marvelapp/adapter/CharacterAdapter.kt [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    override fun getItemCount(): Int = characters.size
+
+    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
+        private val binding = CardviewCharacterBinding.bind(itemView)
+        fun bind(item: Character) {
+            item.let {
+                binding.apply {
+                    Glide.with(itemView.context).load(item.img).into(this.cardviewImage);
```
Comment (SartiMau) [inline app/src/main/java/com/example/marvelapp/adapter/CharacterAdapter.kt]: please put each call in a separated line

Comment (SartiMau) [reply app/src/main/java/com/example/marvelapp/adapter/CharacterAdapter.kt]: and remove de ;

File: app/src/main/java/com/example/marvelapp/util/TransformExtension.kt [hunk 2]
```diff
@@ -6,12 +6,14 @@ import com.example.marvelapp.service.response.DataResponse
 
 fun DataResponse.transformToList(): List<Character> {
     val characterList = mutableListOf<Character>()
-    data.characters.forEach() { characterList.add(Character(it.id, it.name, it.description)) }
+    data.characters.forEach() {
+        characterList.add(Character(it.id, it.name, it.description, (it.img.path+"."+it.img.ext)))
```
Comment (SartiMau) [inline app/src/main/java/com/example/marvelapp/util/TransformExtension.kt]: do not hardcode, please use a DOT constant

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/main/res/drawable/border.xml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
+<shape xmlns:android="http://schemas.android.com/apk/res/android"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent">
+    <solid
+        android:color="@color/black" />
+    <corners
+        android:radius="25dp" />
```
Comment (Georgina-Fernandez) [inline app/src/main/res/drawable/border.xml]: Please, don 't use hardcode. Use dimens

File: app/src/main/res/drawable/card.xml [hunk 4]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="utf-8"?>
+<shape xmlns:android="http://schemas.android.com/apk/res/android"
+    android:shape="rectangle">
+    <gradient
+        android:angle="180"
```
Comment (Georgina-Fernandez) [inline app/src/main/res/drawable/card.xml]: Please don't use hardcodes.

File: app/src/main/res/drawable/card.xml [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+        android:endColor="@color/black"
+        android:startColor="@color/dark_red_text" />
+    <stroke
+        android:width="6dp"
```
Comment (Georgina-Fernandez) [inline app/src/main/res/drawable/card.xml]: Same here

File: app/src/main/res/drawable/card.xml [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+        android:color="#DF000000" />
+    <corners
+        android:bottomLeftRadius="0dp"
```
Comment (Georgina-Fernandez) [inline app/src/main/res/drawable/card.xml]: Same here

File: app/src/main/res/drawable/card.xml [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+        android:radius="75dp"
```
Comment (Georgina-Fernandez) [inline app/src/main/res/drawable/card.xml]: Same here

File: app/src/main/res/drawable/card.xml [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+        android:topLeftRadius="0dp" />
```
Comment (Georgina-Fernandez) [inline app/src/main/res/drawable/card.xml]: Same here

File: app/src/main/res/drawable/card.xml [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+    <padding
+        android:bottom="12dp"
```
Comment (Georgina-Fernandez) [inline app/src/main/res/drawable/card.xml]: Same here

File: app/src/main/res/drawable/card.xml [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+        android:left="22dp"
```
Comment (Georgina-Fernandez) [inline app/src/main/res/drawable/card.xml]: Same here

File: app/src/main/res/drawable/card.xml [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+        android:right="22dp"
```
Comment (Georgina-Fernandez) [inline app/src/main/res/drawable/card.xml]: Same here

File: app/src/main/res/drawable/card.xml [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+        android:top="12dp" />
```
Comment (Georgina-Fernandez) [inline app/src/main/res/drawable/card.xml]: Same here

File: app/src/main/res/layout/activity_characters.xml [hunk 13]
```diff
@@ -33,4 +32,30 @@
         app:lottie_loop="true"
         app:lottie_rawRes="@raw/loading" />
 
+    <TextView
+        android:id="@+id/characters_activity_title"
+        style="@style/app_default_style"
+        android:textSize="32dp"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:layout_marginTop="30dp"
```
Comment (Georgina-Fernandez) [inline app/src/main/res/layout/activity_characters.xml]: Same here

File: app/src/main/res/layout/activity_characters.xml [hunk 14]
```diff
@@ -33,4 +32,30 @@
         app:lottie_loop="true"
         app:lottie_rawRes="@raw/loading" />
 
+    <TextView
+        android:id="@+id/characters_activity_title"
+        style="@style/app_default_style"
+        android:textSize="32dp"
```
Comment (Georgina-Fernandez) [inline app/src/main/res/layout/activity_characters.xml]: Same here

File: app/src/main/res/layout/activity_characters.xml [hunk 15]
```diff
@@ -33,4 +32,30 @@
         app:lottie_loop="true"
         app:lottie_rawRes="@raw/loading" />
 
+    <TextView
+        android:id="@+id/characters_activity_title"
+        style="@style/app_default_style"
```
Comment (Georgina-Fernandez) [inline app/src/main/res/layout/activity_characters.xml:38]: There's some lines here that you may include into a style file. Please check it

File: app/src/main/res/layout/activity_characters.xml [hunk 16]
```diff
@@ -33,4 +32,30 @@
...
+        android:text="@string/characters_activity_title"
+        app:layout_constraintBottom_toTopOf="@id/characters_activity_recyclerview"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent" />
+
+    <androidx.recyclerview.widget.RecyclerView
+        android:id="@+id/characters_activity_recyclerview"
+        android:layout_width="match_parent"
+        android:layout_height="0dp"
```
Comment (Georgina-Fernandez) [inline app/src/main/res/layout/activity_characters.xml]: Please don't use hardcodes

File: app/src/main/res/layout/activity_characters.xml [hunk 17]
```diff
@@ -33,4 +32,30 @@
...
+        android:layout_marginTop="10dp"
```
Comment (Georgina-Fernandez) [inline app/src/main/res/layout/activity_characters.xml]: Same here

File: app/src/main/res/layout/activity_characters.xml [hunk 18]
```diff
@@ -33,4 +32,30 @@
...
+        android:layout_marginRight="10dp"
```
Comment (Georgina-Fernandez) [inline app/src/main/res/layout/activity_characters.xml]: Same here

File: app/src/main/res/layout/cardview_character.xml [hunk 19]
```diff
@@ -0,0 +1,42 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:layout_width="match_parent"
+    android:paddingLeft="10dp"
```
Comment (Georgina-Fernandez) [inline app/src/main/res/layout/cardview_character.xml]: Same here

File: app/src/main/res/layout/cardview_character.xml [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+    android:layout_height="wrap_content"
+    android:layout_marginVertical="15dp"
```
Comment (Georgina-Fernandez) [inline app/src/main/res/layout/cardview_character.xml]: Same here

File: app/src/main/res/layout/cardview_character.xml [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+    android:layout_marginRight="10dp"
+    android:background="@drawable/card">
+
+    <ImageView
+        android:id="@+id/cardview_image"
+        android:layout_width="150dp"
+        android:layout_height="150dp"
```
Comment (Georgina-Fernandez) [inline app/src/main/res/layout/cardview_character.xml]: Same here

File: app/src/main/res/layout/cardview_character.xml [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+    android:layout_width="match_parent"
+    android:paddingLeft="10dp"
+    android:layout_height="wrap_content"
+    android:layout_marginVertical="15dp"
+    android:layout_marginRight="10dp"
+    android:background="@drawable/card">
+
+    <ImageView
+        android:id="@+id/cardview_image"
+        android:layout_width="150dp"
```
Comment (Georgina-Fernandez) [inline app/src/main/res/layout/cardview_character.xml]: Same here

File: app/src/main/res/layout/cardview_character.xml [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+        android:layout_height="150dp"
+        android:layout_marginTop="35dp"
```
Comment (Georgina-Fernandez) [inline app/src/main/res/layout/cardview_character.xml]: Same here

File: app/src/main/res/layout/cardview_character.xml [hunk 24]
```diff
@@ -0,0 +1,42 @@
...
+        android:layout_marginHorizontal="25dp"
```
Comment (Georgina-Fernandez) [inline app/src/main/res/layout/cardview_character.xml]: Same here

File: app/src/main/res/layout/cardview_character.xml [hunk 25]
```diff
@@ -0,0 +1,42 @@
...
+        android:background="@drawable/border"
+        android:padding="14dp"
```
Comment (Georgina-Fernandez) [inline app/src/main/res/layout/cardview_character.xml]: Same here

File: app/src/main/res/layout/cardview_character.xml [hunk 26]
```diff
@@ -0,0 +1,42 @@
...
+        app:layout_constraintBottom_toTopOf="@id/cardview_name"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent" />
+
+    <TextView
+        android:id="@+id/cardview_name"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:text=" CHARACTER NAME "
```
Comment (Georgina-Fernandez) [inline app/src/main/res/layout/cardview_character.xml]: Same here

File: app/src/main/res/layout/cardview_character.xml [hunk 27]
```diff
@@ -0,0 +1,42 @@
...
+        android:textColor="@color/white"
+        android:gravity="center"
+        android:layout_margin="12dp"
```
Comment (Georgina-Fernandez) [inline app/src/main/res/layout/cardview_character.xml]: Same here

File: app/src/main/res/layout/cardview_character.xml [hunk 28]
```diff
@@ -0,0 +1,42 @@
...
+        android:textSize="18sp"
```
Comment (Georgina-Fernandez) [inline app/src/main/res/layout/cardview_character.xml]: Same here

File: app/src/main/res/layout/cardview_character.xml [hunk 29]
```diff
@@ -0,0 +1,42 @@
...
+        android:textStyle="bold"
+        android:textAllCaps="true"
+        android:lines="2"
```
Comment (Georgina-Fernandez) [inline app/src/main/res/layout/cardview_character.xml]: Same here

File: app/src/main/java/com/example/marvelapp/util/TransformExtension.kt (hunk 2 above)
Comment (AndresConfortiGL) [reply app/src/main/java/com/example/marvelapp/util/TransformExtension.kt]: done

File: app/src/main/java/com/example/marvelapp/adapter/CharacterAdapter.kt (hunk 1 above)
Comment (AndresConfortiGL) [reply app/src/main/java/com/example/marvelapp/adapter/CharacterAdapter.kt]: done

**Review by Georgina-Fernandez — APPROVED** [review]

File: app/src/main/java/com/example/marvelapp/util/TransformExtension.kt [hunk 30]
```diff
@@ -4,14 +4,18 @@ import com.example.marvelapp.database.entity.CharacterEntity
 import com.example.marvelapp.entity.Character
 import com.example.marvelapp.service.response.DataResponse
 
+private const val DOT = "."
+
 fun DataResponse.transformToList(): List<Character> {
     val characterList = mutableListOf<Character>()
-    data.characters.forEach() { characterList.add(Character(it.id, it.name, it.description)) }
+    data.characters.forEach() {
+        characterList.add(Character(it.id, it.name, it.description, (it.img.path+DOT+it.img.ext)))
```
Comment (SartiMau) [inline app/src/main/java/com/example/marvelapp/util/TransformExtension.kt]: use a more kotlin style way to concatenate the strings, please

"${it.img.path}$DOT${it.img.ext}"

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/marvelapp/util/TransformExtension.kt (hunk 30 above)
Comment (AndresConfortiGL) [reply app/src/main/java/com/example/marvelapp/util/TransformExtension.kt]: done

**Review by SartiMau — APPROVED** [review]

**Pull request merged**